mapgen/iwyu-mapgen-apple-libc.py

```diff
@@ -0,0 +1,129 @@
+#!/usr/bin/env python3
+
+##===--- iwyu-mapgen-apple-libc.py ----------------------------------------===##
+#
+#                     The LLVM Compiler Infrastructure
+#
+# This file is distributed under the University of Illinois Open Source
+# License. See LICENSE.TXT for details.
+#
+##===----------------------------------------------------------------------===##
+
+""" Generate mappings for Apple Libc.
+
+Map headers that either start with an underscore or are architecture-specific.
+"""
+import argparse
+import json
+import sys
+
+from operator import itemgetter
+from pathlib import Path
+
+
+PATTERN_MAPPINGS = {
+    'pthread_*': 'pthread.h',
+    'sys/_types/_fd_*': 'sys/select.h',
+    'sys/_types/_in_*': 'netinet/in.h',
+    'sys/_types/_o_*': 'fcntl.h',
+    'sys/_types/_s_*': 'sys/stat.h',
+    'sys/_types/_seek_*': 'stdio.h',
+    'sys/ttycom.h': 'sys/ioctl.h',
+}
+
+
+def main(dir, verbose):
+    """ Entry point. """
+
+    root = Path(dir)
+
+    paths = set(root.glob('**/*.h'))
+
+    mappings = []
+
+    for p in paths:
+        initial = p
+
+        # eg. xlocale/_stdio.h -> _stdio.h
+        if p.parent.name == 'xlocale':
+            public_path = p.parent.with_name(p.name)
+            if public_path in paths:
+                p = public_path
+
+        # eg. _stdio.h -> stdio.h
+        while p.name.startswith('_'):
+            public_path = p.with_name(p.name[1:])
+            if public_path in paths:
+                p = public_path
+            else:
+                break
+
+        # arm -> machine
+        if p.parent.name in {'arm', 'arm64', 'i386'}:
+            public_path = p.parent.with_name('machine') / p.name
+            if public_path in paths:
+                p = public_path
+
+        # eg. sys/time.h -> time.h
+        # do not map machine/profile.h to profile.h which is a kerberos header
+        if p.parent.name in {
+            'arm', 'arm64' 'i386', 'machine', 'pthread', 'sys'
+        } and p.name != 'profile.h':
+            public_path = p.parent.with_name(p.name)
+            if public_path in paths:
+                p = public_path
+
+        # architecture/byte_order.h -> machine/byte_order.h
+        if p.parent.name == 'architecture':
+            public_path = p.parent.with_name('machine') / p.name
+            if public_path in paths:
+                p = public_path
+
+        # map other headers
+        for glob, repl in PATTERN_MAPPINGS.items():
+            if p.relative_to(root).full_match(glob):
```

Comment:
I noticed when testing the revised version that `Path.full_match` is not available until Python 3.13, so this fails to run on my fairly recent Ubuntu 24.04 system. Let's see if we can relax that requirement a bit when fixing the quotes issue too?

Reply:
It can be changed to `Path.match` which is less strict but maybe is good enough.